File: cpp/32.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int longestValidParentheses(string s) {
        stack<int> st;
        st.push(-1);
        int n = s.size(), maxlen = 0;
        for (int i = 0; i < n; ++i) {
            if (s[i] == '(') st.push(i);
            else {
                st.pop();
                if (st.empty()) st.push(i);
                else maxlen = max(maxlen, i - st.top());
            }
        }
        return maxlen;
    }
};

File: python/32.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        maxlen = 0
        stack = [-1]
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
            else:
                stack.pop()  # 弹出栈
            if not len(stack):
                stack.append(i)
            else:
                maxlen = max(maxlen, i - stack[-1])

        return maxlen